### PR TITLE
Split View and Shortcuts added

### PR DESCRIPTION
**Pull request by ananay**
* Added Split Panes
* Added 'View' Menu
* Added the following shortcuts:
  * CTRL+1 for Disassembly
  * CTRL+2 for Pseudocode
  * CTRL+3 for Symbolic Execution
  * CTRL+4 for States
  * CTRL+5 for Strings
  * CTRL+Tab for Next Tab
  * CTRL+Shift+Tab for Previous Tab
  * CTRL+D for Split View
  * CTRL+U for Unsplit View

### CODE REVIEW

File: angrmanagement/ui/main_window.py [hunk 1]
```diff
@@ -159,30 +167,68 @@ def _init_toolbars(self):
...
     #
 
+    def _tabify(self):
+        self.right_dockable_views = [dock for dock in           self.workspace.dockable_views if dock.widget() is not None and dock.widget().default_docking_position == 'right']
+
+        for d0, d1 in zip(self.right_dockable_views, self.right_dockable_views[1:]):
+            self.central_widget.tabifyDockWidget(d0, d1)
+        self.right_dockable_views[0].raise_()
+
+
```
Comment (sraboy) [inline angrmanagement/ui/main_window.py]: Only one line between member functions, two for module funcs.

File: angrmanagement/ui/workspace.py [hunk 2]
```diff
@@ -85,6 +92,28 @@ def on_function_tagged(self):
     # Public methods
     #
 
+
```
Comment (sraboy) [inline angrmanagement/ui/workspace.py]: Extra line break. Here and elsewhere.

File: angrmanagement/ui/workspace.py [hunk 3]
```diff
@@ -108,6 +137,58 @@ def add_view(self, view, caption, category):
...
+            dock.setWidget(self.default_tabs[id])
+            
+            self.dockable_views[id] = dock
+            self._main_window.central_widget_main.setStretchFactor(1,0)
+            self._main_window._tabify()
+            self.is_split = 0
+    
+
+    def add_view2(self, view, caption, category):
```
Comment (sraboy) [inline angrmanagement/ui/workspace.py]: `add_view2` doesn't seem referenced anywhere.

**Comment by sraboy** [comment]
My (unsolicited) review comments notwithstanding, this is a very welcome feature.

File: angrmanagement/ui/workspace.py [hunk 4]
```diff
@@ -85,6 +92,28 @@ def on_function_tagged(self):
     # Public methods
     #
 
+
+    def split_view(self):
+        id = self._main_window.getCurrentTabId()
+        if self.is_split == 0:
+            print("Split view called")
```
Comment (sraboy) [inline angrmanagement/ui/workspace.py]: I'm guessing you forgot this.

File: angrmanagement/ui/workspace.py [hunk 5]
```diff
@@ -108,6 +137,58 @@ def add_view(self, view, caption, category):
         self.dockable_views.append(dock)
         self.view_to_dockable[view] = dock
 
+
+    def unsplit_view(self):
+        if self.is_split == 1:
+            id = self.split_tab_id
+            print("Unsplit view called")
```
Comment (sraboy) [inline angrmanagement/ui/workspace.py]: This one too.

File: angrmanagement/ui/main_window.py [hunk 6]
```diff
@@ -159,30 +167,68 @@ def _init_toolbars(self):
...
         self.menuBar().addMenu(analyzeMenu.qmenu())
+        self.menuBar().addMenu(viewMenu.qmenu())
         self.menuBar().addMenu(helpMenu.qmenu())
 
     #
     # Workspace
     #
 
+    def _tabify(self):
+        self.right_dockable_views = [dock for dock in           self.workspace.dockable_views if dock.widget() is not None and dock.widget().default_docking_position == 'right']
```
Comment (sraboy) [inline angrmanagement/ui/main_window.py]: Extra spaces after `in`.

**Comment by ananay** [comment]
@sraboy Thank you for your review, I will make the requested changes and push more commits to this PR.

File: angrmanagement/ui/main_window.py [hunk 7]
```diff
@@ -159,30 +167,68 @@ def _init_toolbars(self):
...
+        self.central_widget.setTabPosition(Qt.RightDockWidgetArea, QTabWidget.North)
 
-        for d0, d1 in zip(right_dockable_views, right_dockable_views[1:]):
-            self.central_widget.tabifyDockWidget(d0, d1)
-        right_dockable_views[0].raise_()
 
-        self.central_widget.setTabPosition(Qt.RightDockWidgetArea, QTabWidget.North)
+    def getCurrentTabId(self):
+        for i in range(1,6):
+            if self.right_dockable_views[i-1].visibleRegion().isEmpty() == False:
```
Comment (sraboy) [inline angrmanagement/ui/main_window.py]: I think `if not ...` is less misleading. Also, `getCurrentTabId` could just be a `current_tab` property... either way, prefer underscores over CamelCase.

File: angrmanagement/ui/main_window.py [hunk 8]
```diff
@@ -159,30 +167,68 @@ def _init_toolbars(self):
...
+                return i
+        return 1
+
+    #
+    # Tab switching
+    #
+
+    def nextTab(self):
```
Comment (sraboy) [inline angrmanagement/ui/main_window.py]: Generally, prefer `underscore_func_name()` over CamelCase. Here and elsewhere.

File: README.md [hunk 9]
```diff
@@ -40,9 +40,12 @@ git clone git@github.com:angr/angr-management.git
 ```
 cd angr-management
 pip install -e .
+pip install -r requirements.txt
```
Comment (rhelmot) [inline README.md]: This isn't necessary. It may have been before today but I fixed a typo in setup.py.

**Comment by rhelmot** [comment]
This looks good from a code review perspective. Someone (me perhaps!) should pull it down and see how it performs after the issues are addressed.

**Comment by ananay** [comment]
All changes done!

File: angrmanagement/ui/workspace.py [hunk 10]
```diff
@@ -108,6 +137,33 @@ def add_view(self, view, caption, category):
...
+            docking_positions = {
+                'left': Qt.LeftDockWidgetArea,
+                'right': Qt.RightDockWidgetArea,
+                'top': Qt.TopDockWidgetArea,
+                'bottom': Qt.BottomDockWidgetArea,
+            }
+        
+            self._main_window.central_widget2.removeDockWidget(self.last_unsplit_view)
+
+
```
Comment (ltfish) [inline angrmanagement/ui/workspace.py]: Nitpick: Too many blank lines. One is enough. Also, please move `unsplit_view()` into the new view manager (PR #73).

File: angrmanagement/ui/workspace.py [hunk 11]
```diff
@@ -108,6 +137,33 @@ def add_view(self, view, caption, category):
         self.dockable_views.append(dock)
         self.view_to_dockable[view] = dock
 
+    def unsplit_view(self):
+        if self.is_split == 1:
+            id = self.split_tab_id
```
Comment (ltfish) [inline angrmanagement/ui/workspace.py]: Nitpick: Do not shadow the builtin function `id`. Use a different variable name.

File: angrmanagement/ui/workspace.py [hunk 12]
```diff
@@ -85,6 +94,26 @@ def on_function_tagged(self):
     # Public methods
     #
 
+    def split_view(self):
+        id = self._main_window.getCurrentTabId()
```
Comment (ltfish) [inline angrmanagement/ui/workspace.py]: Same here: Do not shadow `id`. Move `split_view()` to the new view manager.

File: angrmanagement/ui/workspace.py [hunk 13]
```diff
@@ -25,14 +28,20 @@ def __init__(self, main_window, instance):
...
+        self.splitter_state = QSettings()
+        self.splitter_state.setValue("splitterSizes", self._main_window.central_widget_main.saveState())
 
         #
         # Initialize font configurations
         #
         Conf.init_font_config()
 
-        default_tabs = [
+        self.default_tabs = [
```
Comment (ltfish) [inline angrmanagement/ui/workspace.py]: What's the rationale behind this change?

Comment (ltfish) [reply angrmanagement/ui/workspace.py]: I forgot to add that most (if not all) default tabs can be closed by user as needed. So we should use the new view manager to get a reference to the currently open view.

File: angrmanagement/ui/workspace.py [hunk 14]
```diff
@@ -2,7 +2,10 @@
 import logging
 from collections import defaultdict
 
-from PySide2.QtCore import Qt
+from PySide2.QtCore import Qt, QSettings
+
+from PySide2.QtWidgets import QSplitter, QMainWindow, QShortcut
```
Comment (ltfish) [inline angrmanagement/ui/workspace.py]: Nitpick: The blank line above this import is unnecessary. All imports from PySide2 should be grouped together.

File: angrmanagement/ui/menus/view_menu.py [hunk 15]
```diff
@@ -0,0 +1,17 @@
+from PySide2.QtGui import QKeySequence
+from PySide2.QtCore import Qt
+
+from .menu import Menu, MenuEntry, MenuSeparator
+
+
+class ViewMenu(Menu):
+    def __init__(self, main_window):
```
Comment (ltfish) [inline angrmanagement/ui/menus/view_menu.py]: Awesome! I've thinking of adding a view menu for a while.

File: angrmanagement/ui/main_window.py (hunk 7 above)
Comment (ltfish) [reply angrmanagement/ui/main_window.py]: Please use `xxx is False` if you are comparing something with False.

File: angrmanagement/ui/main_window.py [hunk 16]
```diff
@@ -13,11 +13,13 @@
 from .menus.file_menu import FileMenu
 from .menus.analyze_menu import AnalyzeMenu
 from .menus.help_menu import HelpMenu
+from .menus.view_menu import ViewMenu
 from ..config import IMG_LOCATION
 from .workspace import Workspace
 from .dialogs.load_binary import LoadBinary, LoadBinaryError
 from .dialogs.new_state import NewState
 from .toolbars import StatesToolbar, AnalysisToolbar, FileToolbar
+from functools import partial
```
Comment (ltfish) [inline angrmanagement/ui/main_window.py]: Nitpick: Please move this import line to the first group of import (together with `import pickle` and `import os`) since `functools` is a Python default package, just like `pickle` and `os`.

File: README.md [hunk 17]
```diff
@@ -56,6 +58,18 @@ Or if you have a development install:
 python start.py
 ```
 
+### Shortcuts
```
Comment (ltfish) [inline README.md]: Good job in documenting your improvements!

**Comment by ltfish** [comment]
Thank you. I added a few comments. Please first rebase your code to the latest master (since we merged in the view manager yesterday) and then address the left-over comments. I'll click merge after that. Thanks a lot!

Note to myself: CI for angr management is a must at this moment...

**Comment by ananay** [comment]
@ltfish @sraboy Done!
As @sraboy suggested, I've made CTRL+D to toggle the split / unsplit view feature.

**Comment by sraboy** [comment]
Just got back from traveling but I'll test this again in the next few days.

**Comment by ananay** [comment]
@sraboy @ltfish following up on this

**Comment by sraboy** [comment]
I haven't had a chance to check out the code itself but Ctrl-4 doesn't go to `States`. Also, Previous/Next tab doesn't work when the window is split:

```
Traceback (most recent call last):
  File "/home/steve/code/angr-dev/angr-management/angrmanagement/ui/view_manager.py", line 126, in next_tab
    right_dockable_views = self.get_right_views()
  File "/home/steve/code/angr-dev/angr-management/angrmanagement/ui/view_manager.py", line 80, in get_right_views
    return [dock for dock in self.docks if dock.widget().default_docking_position == 'right']
  File "/home/steve/code/angr-dev/angr-management/angrmanagement/ui/view_manager.py", line 80, in <listcomp>
    return [dock for dock in self.docks if dock.widget().default_docking_position == 'right']
AttributeError: 'NoneType' object has no attribute 'default_docking_position'
```
and previous tab:
```
Traceback (most recent call last):
  File "/home/steve/code/angr-dev/angr-management/angrmanagement/ui/view_manager.py", line 136, in previous_tab
    right_dockable_views = self.get_right_views()
  File "/home/steve/code/angr-dev/angr-management/angrmanagement/ui/view_manager.py", line 80, in get_right_views
    return [dock for dock in self.docks if dock.widget().default_docking_position == 'right']
  File "/home/steve/code/angr-dev/angr-management/angrmanagement/ui/view_manager.py", line 80, in <listcomp>
    return [dock for dock in self.docks if dock.widget().default_docking_position == 'right']
AttributeError: 'NoneType' object has no attribute 'default_docking_position'
```

**Comment by ananay** [comment]
@sraboy fixed! Please let me know if you find any other issue

**Comment by sraboy** [comment]
Awesome! One last bug I found is that next tab doesn't wrap around but previous tab does:.

```
Traceback (most recent call last):
  File "/home/steve/code/angr-dev/angr-management/angrmanagement/ui/view_manager.py", line 133, in next_tab
    right_dockable_views[self.get_current_tab_id()].raise_()
IndexError: list index out of range
```

I noticed keyboard focus doesn't follow the next/prev tab. Imagine I have the `Dissassembly` view on the left and split out the `Interaction` to the right. I type some things in there. Now I want to use the `g` hotkey to jump to a new address in the `Disassembly`. I can't do that without using the mouse to click on the `Disassembly` pane first and then use the mouse to click back to the `Interaction` split pane. Then again, this could be a "feature" if someone is interacting and merely wants to change from `Disassembly` to `Strings` without messing with their keyboard focus; maybe this is good as-is for splitting and keyboard focus would only update if we do a more traditional MDI-tab-split interface later... I dunno. @ltfish? 

If you do change tabbing to include keyboard focus, it would be nice if next/prev tab also cycled keyboard focus through/to the split pane too.

**Comment by ananay** [comment]
@sraboy Added the wrap for next tab as well!

File: angrmanagement/ui/view_manager.py [hunk 18]
```diff
@@ -71,6 +70,22 @@ def raise_view(self, view):
...
+        """
+
+        docks = []
+        for dock in self.docks:
+            if dock.widget() is not None:
+                if dock.widget().default_docking_position == 'right':
+                    docks.append(dock)
+        return docks
+
+        #return [dock for dock in self.docks if dock.widget().default_docking_position == 'right']
```
Comment (ltfish) [inline angrmanagement/ui/view_manager.py]: Please remove this line of comment if it's not going to be used in the future.

File: README.md [hunk 19]
```diff
@@ -56,6 +58,17 @@ Or if you have a development install:
...
+- Load a new binary: ```Ctrl + O```
```
Comment (zwimer) [inline README.md]: Add load a new docker target ctrl + shift + o

File: angrmanagement/ui/main_window.py [hunk 20]
```diff
@@ -182,24 +187,53 @@ def _init_toolbars(self):
...
+
+    def _init_shortcuts(self):
+        """
+        Initialize shortcuts
+
+        :return:    None
+        """
+
+        right_dockable_views = self.workspace.view_manager.get_right_views()
+        for i in range(1,7):
```
Comment (zwimer) [inline angrmanagement/ui/main_window.py]: Avoid hardcoding the number of tabs. It makes things easier to modify

File: angrmanagement/ui/view_manager.py [hunk 21]
```diff
@@ -90,9 +105,44 @@ def tabify_right_views(self):
...
+
+    def get_current_tab_id(self):
+        """
+        Get Current Tab ID
+        
+        :return:    Tab ID (int)
+        """
+        
+        right_dockable_views = self.get_right_views()
+        for i in range(1,7):
```
Comment (zwimer) [inline angrmanagement/ui/view_manager.py]: Hard coding again

**Comment by ltfish** [comment]
Thanks! Looks good to me now. I'm ready to merge it if no one objects.

File: angrmanagement/ui/workspace.py (hunk 13 above)
Comment (ananay) [reply angrmanagement/ui/workspace.py]: I added self so it can be accessed by other methods

Comment (ltfish) [reply angrmanagement/ui/workspace.py]: Gotcha. I missed that. Thanks.

File: angrmanagement/ui/workspace.py (hunk 14 above)
Comment (ananay) [reply angrmanagement/ui/workspace.py]: Oops, sorry. I missed this one. Fixed.

**Comment by ltfish** [comment]
Merged. Thanks!
